HW0/P5/P5.txt

```diff
@@ -1 +1,86 @@
+
+
+1 bag to 1 cashier in 1s:
+If we can only transfer one bag to one cashier in one s, we will cause cashiers to wait till enough data/bags have arrived to perform the actual summation. This bottleneck actually causes, that some cashiers are not needed as we cannot provide enough cashiers bags for simultaneous computation.
+
+Thus, we can imagine the workflow as following:
+(Let c1 be cashier1, c2 cashier 2, ...)
+
+startphase:
+Give 1 bag to c1 | c2 idle | 1s
+Give 1 bag to c1 | c2 idle | 1s
+c1 sums | Give 1 bag to c2 | 1s
+c1 idle | Give 1 bag to c2 | 1s
+
+repeating part:
+Give 1 bag to c1 | c2 sums | 1s
+c1 sums | Give 1 bag to c2 | 1s
+....
+
+endpase:
+c1 idle | c2 tells c1 number | 1s
+c1 sums | c2 idle | 1s
+
+
+Having the capability to provide in one s only one cashier with bags will cause that we are able to use 2 cashiers only. Thus, we will need for 256 bags 256s+2s = 258s compared to 256 * 2s = 512s if we use 1 cashier only (we assume a cashier can only handle one action per time, i.e. it is not possible for him to concurrently sum & receive a bag, cashiers have no multitasking ability).
+
+
+2 bags to one person:
+In this case we can make better use of concurreny. The idea is to strip the data and provide nodes with enough data for computation.
+Using 3 cashiers we can avoid idle (except for start / end):
+
+start phase:
+Give 2 bags to c1 	| c2 idle 				| c3 idle 				| 1s
+c1 sums 			| Give 2 bags to c2 	| c3 idle 				| 1s
+Give 2 bags to c1 	| c2 sums 				| c3 idle 				| 1s
+c1 sums, 1 bag left | Give 2 bags to c2 	| c3 idle 				| 1s
+c1 sums 			| c2 sums, 1 bag left	| Give 2 bags to c3 	| 1s
+Give 2 bags to c1	| c2 sums 				| c3 sums 				| 1s
+c1 sums, 1 bag left | Give 2 bags to c2		| c3 idle				| 1s
+c1 sums 			| c2 sums, 1 bag left   | Give 2 bags to c3 	| 1s
+
+repeating phase:
+Give 2 bags to c1 	| c2 sums 				| c3 sums, 1 bag left 	| 1s
+c1 sums, 1 bag left	| Give 2 bags to c2		| c3 sums 				| 1s
+c1 sums 			| c2 sums, 1 bag left 	| Give 2 bags to c3 	| 1s
+...
+
+endphase:
+c1 send to c3		| c2 sums, 1 bag left 	| c3 receives		 	| 1s
+c1 idle				| c2 sums 				| c3 sums 				| 1s
+c1 idle 			| c2 send to c3 		| c3 receives 			| 1s
+c1 idle 			| c2 idle 				| c3 sums 				| 1s
+
+==> time needed to compute counter is now only 8s (16 bags processed) + 3s * 40 (40 * 6 bags = 240 bags = 256 bags - 16 bags) + 4s (combine) 
+==> 132s
+
+However, it is not possible to use (in a meaningful way) more than 3 nodes/employees. Why? We know that each node can process in 1s 2 numbers. But we are also only able to provide in 1s 2 bags. Thus our "bandwidth" of providing data is the same as the "bandwidth" of a node in terms of what he can process. In order to use more nodes and speedup the whole task, we will need the ability to give more than 2 bags to a cashier or in other terms a network/cluster cannot process more than the bandwidth of the data source. 
```

Comment:
What about exponential distribution?  I.e. have employees also distribute bags.  This can be quite fast